Add basic support for PEP 702 (@deprecated).

tyralla opened the pull request:
Closes #16111

This PR provides only basic support.  Many special cases might need additional attention (descriptors, some special methods like `__int__`, etc.).  Other open issues are code comments, eventual documentation updates, the deprecation message style, etc.).  But I wanted to offer these first steps before going on vacation (so I cannot respond to possible reviews too soon).

Maybe someone wants to extend the list of (test) cases the basic support should address?

JelleZijlstra commented:
Thank you! I'll review this soon.

TeamSpen210 commented on mypy/checker.py:
```diff
@@ -7519,6 +7556,55 @@ def has_valid_attribute(self, typ: Type, name: str) -> bool:
+                    (callee.fullname in DEPRECATED_TYPE_NAMES)
+                    or (not callee.fullname and callee.name == "deprecated")
+                )
+                and (len(args := decorator.args) == 1)
```
This should be `>=`, since `deprecated()` takes some keyword args too for runtime warning behaviour. MyPy doesn't need to do anything special with those though.

tyralla replied:
Thanks, done

ilevkivskyi commented on mypy/errorcodes.py:
```diff
@@ -293,5 +293,11 @@ def __hash__(self) -> int:
+DEPRECATED: Final = ErrorCode(
+    "deprecated",
+    "Warn when importing or using deprecated (overloaded) functions, methods or classes",
+    "General",
```
I think this should be `default_enabled=False` (at least for now). Also you must update the docs to include this error code (this is why the docs build failed).

JelleZijlstra replied:
When you say "for now", what kind of timeline do you have in mind for turning it on by default? I see that it will be somewhat painful for users who now get these deprecation warnings, but that will also be the case if we turn it on in the future.

It might be nice to present deprecations as warnings by default (i.e., they don't break the type check), but provide an option for users who want any use of deprecated functionality to fail CI.

ilevkivskyi replied:
If we turn it on in the future people will be at least somewhat ready.

Mypy doesn't have warnings, it is either an `error` or `note`. That said, `note` may be fine (since notes result in exit code 0), but also IIRC there is no way make a specific note an error.

tyralla replied:
I switched from `error` to `note`.  I can still pass the new error code, which makes a difference when using `type: ignore[deprecated]` or other error codes.  However, the error code does not seem to be appended to the message.

ilevkivskyi replied:
@tyralla Yes, error codes on notes are not shown by default, you need to add it to `SHOW_NOTE_CODES` in `errors.py`.

tyralla replied:
Thanks, done. I added three tests which cover the different modes (no reports, notes, errors).

JelleZijlstra commented on mypy/checker.py:
```diff
@@ -654,6 +659,16 @@ def _visit_overloaded_func_def(self, defn: OverloadedFuncDef) -> None:
         if defn.impl:
             defn.impl.accept(self)
+            if (
+                isinstance(impl := defn.impl, Decorator)
```
I feel this is too many walruses, especially with the `ct` variable name being reused.

tyralla replied:
I agree. Only one walrus left.

JelleZijlstra commented on mypy/checker.py:
```diff
@@ -7519,6 +7556,55 @@ def has_valid_attribute(self, typ: Type, name: str) -> bool:
+                and isinstance(callee := decorator.callee, NameExpr)
+                and (
+                    (callee.fullname in DEPRECATED_TYPE_NAMES)
+                    or (not callee.fullname and callee.name == "deprecated")
```
Does this mean that an unresolved decorator `@deprecated` gets treated like the PEP 702 deprecated?

tyralla replied:
Yes, I am afraid so.  

I wrote this line when working on `testDeprecatedProperty`.  In `TypeChecker._visit_overloaded_func_def`, `defn.items[1].callee.fullname == ""`.  Is this a Mypy bug or am I missing something?

I will remove this line and let the test fail for now.

tyralla replied:
It seems the semantic analyser needed a little tweak to analyse the additional decorators of the non-getter overload parts.   `testDeprecatedProperty` passes again.  Other decorators named `deprecated` should not become confused with those of  `warnings` and `typing_extensions` anymore.

JelleZijlstra commented on test-data/unit/check-deprecated.test:
```diff
@@ -0,0 +1,353 @@
+f[1]  # E: __main__.f is deprecated - use f2 instead \
+      # E: Value of type "Callable[[], None]" is not indexable
+g = f  # E: __main__.f is deprecated - use f2 instead
+g()  # E: __main__.f is deprecated - use f2 instead
```
This ideally shouldn't error since we already got an error at the `g = f` line. I don't mind if that's too hard to implement though.

tyralla replied:
Unfortunately, I have no spontaneous idea how to implement this with reasonable effort.

tyralla replied:
@JelleZijlstra: Fixed by commit 49ee7a8ee53907b8b65665adcc756e811780811b.

JelleZijlstra commented on mypy/checker.py:
```diff
@@ -7519,6 +7556,55 @@ def has_valid_attribute(self, typ: Type, name: str) -> bool:
+                name = func.fullname
+        else:
+            name = type_.fullname
+        self.msg.fail(f"{name} is deprecated - {deprecated}", context, code=codes.DEPRECATED)
```
A dash doesn't look great here. For reference, in pyanalyze I use a colon:

```
Literal[<function show at 0x105723ba0>] is deprecated: Use 'print' instead (code: deprecated)
```

Pyright gives the user message on a separate line:

```
  /Users/jelle/Desktop/depr.py:10:1 - error: The function "show" is deprecated
    Use 'print' instead (reportDeprecated)
```

I think the colon version works best with mypy's general style of outputting messages.
```suggestion
        self.msg.fail(f"{name} is deprecated: {deprecated}", context, code=codes.DEPRECATED)
```

tyralla replied:
okay, `warn_deprecated` now writes ": " instead of " - "

JelleZijlstra reviewed (commented):
Thanks, this is looking pretty good!

JelleZijlstra commented:
```
mkosi (https://github.com/systemd/mkosi)
+ mkosi/log.py:92:34: error: logging.getLevelName is deprecated - The str -> int case is considered a mistake.  [deprecated]
```

This is a case where just one overload is deprecated, so I think `logging.getLevelName is deprecated` is misleading.

For reference, pyright has the same issue:

```
  /Users/jelle/py/tmp/deproverl.py:18:1 - error: The function "f" is deprecated
    no more floats (reportDeprecated)
```

Pyanalyze prints the affected overload (but appears to have a bug where the signature degenerates to `(*Any, **Any) -> Any`:

```
Use of deprecated overload (*args: Any[inference], **kwargs: Any[inference]) -> Any[unannotated]: no more floats (code: deprecated)
In /Users/jelle/py/tmp/deproverl.py at line 18
```

I would want it to either print the actual deprecated signature (like pyanalyze) or say something like "Deprecated call to logging.getLevelName", indicating that the function isn't fully deprecated, just this way of calling it is deprecated.

ilevkivskyi commented on docs/source/error_code_list2.rst:
```diff
@@ -231,6 +231,40 @@ incorrect control flow or conditional checks that are accidentally always true o
+Check that imported or used feature is deprecated [deprecated]
+--------------------------------------------------------------
+
+If you use :option:`--warn-deprecated <mypy --warn-deprecated>`, mypy generates a note if
```
It looks like there has been some back and forth so docs are out of sync now. To summarize how I see the logic should be:
* Deprecation messages are always shown as notes by default (with code shown, see my previous comment), this will result in exit code 0 (which is kind of my main concern, I don't want users CI to break unnecessarily on new mypy version).
* People who still don't want to see the notes, can use `--disable-error-code=deprecated` and/or `# type: ignore[deprecated]`.
* There should be a flag that *turns notes into errors* (for more pedantic people). They will still be able to selectively silence the errors using `# type: ignore[deprecated]`.

tyralla replied:
> * There should be a flag that _turns notes into errors_ (for more pedantic people).

Does Mypy implement other checks that can result in errors or notes, depending on an option? (I'm just asking to avoid reinventing the wheel.)

ilevkivskyi replied:
TBH I don't remember. But in any case it should be as simple as changing last line in your `warn_deprecated()` method in `checker.py` to something like
```python
if self.options.error_on_deprecated:  # or other flag name
    self.error(...)
else:
    self.note(...)
```

tyralla replied:
That seems doable, thanks.  I will implement it later.

tyralla replied:
Done!  `report-deprecated-as-error` is the shortest understandable description I could come up with.  Other suggestions are welcome, of course.

ilevkivskyi commented on docs/source/error_code_list2.rst:
```diff
@@ -231,6 +231,44 @@ incorrect control flow or conditional checks that are accidentally always true o
+--------------------------------------------------------------
+
+By default, mypy generates a note if your code imports a deprecated feature explicitly with a
+``from mod import depr" statement or uses a deprecated feature imported otherwise or defined
```
```suggestion
``from mod import depr`` statement or uses a deprecated feature imported otherwise or defined
```

tyralla commented:
> I would want it to either print the actual deprecated signature (like pyanalyze) or say something like "Deprecated call to logging.getLevelName", indicating that the function isn't fully deprecated, just this way of calling it is deprecated.

This request was not too hard to implement (at least concerning the available test cases).  The suggested pattern is: 
`__main__.A.__add__ is deprecated [overload def (__main__.A, builtins.int)]: no A + int`

This step offered the chance to report wrong orders of `@deprecated` and `@overload`.  I am not aware this causes any real trouble, so I selected to report the wrong order as a note:
`@overload should be placed before @deprecated`

tyralla commented:
I refactored the code and polished the messages a little.  In my opinion, the "basic support" is ready now.  I may find time to react to possible further remarks today, but then I am off for a while.  So, do not hesitate to make any changes you think are useful by yourself.

ilevkivskyi reviewed (commented):
OK, I had time to look at the implementation, and I think there are some large-scale things that are not right:
* Being deprecated is not a property of a type, it is a property of a symbol, and as such should be only ever stored on symbol nodes. (I understand this may complicate a bit the situation with specific overload deprecated, but this does not justify breaking the abstraction)
* Creating/storing the deprecation warnings should be done during semantic analysis phase (emitting the actual warnings should be kept in type-checking however). Doing this during type-checking will lead to unpredictable results if e.g. an import cycle is involved (I understand this is an edge case, but it is non-fixable in principle with the current logic).
* We need good test coverage for both coarse- and fine-grained incremental modes. It is important that when deprecation status of a symbol changes, then all its (transitive) use sites are re-checked appropriately.

Youssefares commented:
@tyralla are you looking at implementing the feedback for this? I am happy to pick up where you left off otherwise.

tyralla commented:
Sorry, I was too busy the last few weeks to continue working on it. 

@Youssefares: I could try to thoroughly address the last remarks in mid-September, but definitely not earlier, and Python 3.13 is on its way.  So please do not hesitate if you can start sooner.

@ilevkivskyi: 

- As CallableType already has "symbol properties" ('name', 'definition') for error messages, I assumed that adding deprecation messages would be fine.  I have had no time so far to check how many complications moving this property would cause.  (Maybe not only for overloads but also for descriptors and special methods like `__add__`?)
- I also assumed there should be no problem with caching because a relevant change like adding or removing a descriptor should not remain unnoticed (maybe except for changing a deprecation message).  But as you were in doubt, I added a few preliminary fine-grained incremental tests and, in fact, found a thing to improve (8c0260e8b4f529fa9c84cf648c815cf8bcc20e16, just pushed).  Unfortunately, one test case  (with transitive use sites) still fails (cf2dcaf41ddff5a5df2777911675d4bc3f1206f4), and I did not have enough time to investigate why.  All I can say is that when I set up `testAddFunctionDeprecationIndirectImport1` as an actual project (files on disk) and use the daemon "normally", everything seems to work well.  Hence, I am unsure whether this is caused by a flaw in my implementation or a test setup problem.
- I must admit, I still do not know what you mean by "coarse-grained incremental mode".  The "normal" caching on disk when using Mypy repeatedly without the Mypy server being involved?

tyralla commented:
@Youssefares: Did you find time to work on it?  I could restart soon, but likely only in small steps.  (Especially if I have to solve the mystery about `testAddFunctionDeprecationIndirectImport1` myself.)  So, if you think you can finish this PR faster, please do.  Continuing as a Tag Team would also be cool if you like.

tyralla commented:
@Youssefares: Okay,  then I'll pick up again.

@ilevkivskyi: a6d0e59e29523f0493a4a7433c42fd173d262ea7 addresses your second remark (moving the deprecation note creation from `checker.py` to `semanal.py`) without any functional changes, as far as the tests tell.  `testAddFunctionDeprecationIndirectImport1 ` still fails.

tyralla commented:
@ilevkivskyi

I moved the attribute `deprecated` from the types `CallableType and `Overloaded` to the symbol nodes `FuncDef` and `OverloadedFuncDef`.  All in all, it improves the code.  Except for the case of specific overload deprecations.  Maybe you know a better solution than the one I implemented.

After all these changes, ` testAddFunctionDeprecationIndirectImport1` still fails, what strenghtens my impression the the test itself is flawed.

ilevkivskyi commented on mypy/checker.py:
```diff
@@ -2923,6 +2926,16 @@ def visit_assignment_stmt(self, s: AssignmentStmt) -> None:
+                if (
+                    isinstance(lvalue, NameExpr)
+                    and isinstance(var := lvalue.node, Var)
+                    and isinstance(instance := get_proper_type(var.type), Instance)
```
It seems to me this will miss cases where a deprecated class appears in a nested position, e.g. `foo: Deprecated | Other` or `bar: list[Deprecated]`. No need to fix in this PR, you can open a follow up PR, or issue.

ilevkivskyi commented on mypy/checker.py:
```diff
@@ -7519,6 +7542,29 @@ def has_valid_attribute(self, typ: Type, name: str) -> bool:
     def get_expression_type(self, node: Expression, type_context: Type | None = None) -> Type:
         return self.expr_checker.accept(node, type_context=type_context)
 
+    def check_deprecated(self, typ: SymbolNode | None, context: Context) -> None:
```
The parameter name `typ` is now a bit misleading now.

tyralla replied:
Yes, definitively misleading. Changed it to `node`.

ilevkivskyi commented on mypy/checker.py:
```diff
@@ -7519,6 +7542,29 @@ def has_valid_attribute(self, typ: Type, name: str) -> bool:
+            else:
+                self.warn_deprecated(typ, context)
+
+    def warn_deprecated(self, typ: SymbolNode | None, context: Context) -> None:
```
The parameter name `typ` is now a bit misleading now here as well (I will not add this comment, if there are more similar cases).

tyralla replied:
Also changed. I think there are no other similar cases - I added only two `..._deprecated` methods.

ilevkivskyi commented on mypy/semanal.py:
```diff
@@ -1255,10 +1256,51 @@ def analyze_overloaded_func_def(self, defn: OverloadedFuncDef) -> None:
+                        self.msg.note("@overload should be placed before @deprecated", d)
+                    elif (deprecated := self.get_deprecated(d)) is not None:
+                        deprecation = True
+                        if isinstance(type_ := item.func.type, CallableType):
```
We don't use trailing `_`, maybe use `typ` if it is not taken, or like `callable_type` otherwise.

tyralla replied:
Okay, `typ` was still available.

ilevkivskyi reviewed (approved):
> Except for the case of specific overload deprecations

I actually expected this. I think it is a worthy trade-off.
> what strenghtens my impression the the test itself is flawed.

Hm, I don't see anything obviously wrong with the test. As a guess you need to update this part to include the new flag https://github.com/python/mypy/blob/master/mypy/server/astdiff.py#L228-L256 (since currently you effectively rely on node kind switching from function to decorator to trigger the update).

Anyway, my current comments are all minor, this is good to merge as soon as you fix the test.

tyralla commented:
> Hm, I don't see anything obviously wrong with the test.  As a guess, you need to update this part to include the new flag https://github.com/python/mypy/blob/master/mypy/server/astdiff.py#L228-L256 (since currently you effectively rely on node kind switching from function to decorator to trigger the update).

My first attempt to include the new flag into the returned tuple made no difference.  While thinking about it, I came across the `snapshot_symbol_table`.  After the change proposed by commit 09a53d5b9a4663ed924204e344fdfc2bd34eae0a, all tests passed.  However, I have no idea about the further implications of this change (performance?).  I could investigate further next week, but maybe it is directly clear to you whether the proposed change is a good or bad idea.

tyralla commented:
Hm, [docstring of module `astdiff`](https://github.com/python/mypy/blob/1a2c8e2a4df21532e4952191cad74ae50083f4ad/mypy/server/astdiff.py#L47-L50) seems to suggest it is a bad idea.

tyralla commented:
@ilevkivskyi 

I am now quite sure that `testDeprecatedAddFunctionDeprecationIndirectImport-only_when_nocache` fails due to a more general problem (or my misunderstanding of `only_when_nocache`).   Mypy does not report any of the two obvious errors, no matter which of the four decorator definitions I select:

```python
[case testReplaceFunctionWithDecoratedFunction-only_when_nocache]
from b import f
x: int = f()
import b
y: int = b.f()

[file b.py]
from a import f

[file a.py]
def f() -> int: ...

[file a.py.2]
from typing import Callable, TypeVar

T = TypeVar("T")

# def d(t: str, /) -> Callable[[T], T]: ...  # version 1
# def d(t: str, /) -> Callable[[Callable[[], str]], Callable[[], str]]: ...  # version 2
# def d(t: T, /) -> T: ...  # version 3
def d(t: Callable[[], str], /) -> Callable[[], str]: ...  # version 4

# @d("deprecated")  # version 1 or 2
@d  # version 3 or 4
def f() -> str: ...

[builtins fixtures/tuple.pyi]
[out]
==
main:2: error: Incompatible types in assignment (expression has type "str", variable has type "int")
main:4: error: Incompatible types in assignment (expression has type "str", variable has type "int")
```

tyralla commented:
I now had time for some debugging but did not come to a helpful conclusion.   

[The already mentioned docstring](https://github.com/python/mypy/blob/1a2c8e2a4df21532e4952191cad74ae50083f4ad/mypy/server/astdiff.py#L47-L50) seems to suggest that snapshots and fine-grained dependencies have nothing to do with each other.  I thought that snapshots would be used to determine the triggers that are used in combination with the fine-grained dependency mappings for determining targets that need reprocessing.

Another confusion comes from the fact that Mypy does **not** fail the following test cast:

```python
[case testReplaceFunctionWithDecoratedFunction-only_when_nocache]
from b import f
x: int = f()

[file b.py]
from a import f

[file a.py]
def f() -> int: ...

[file a.py.2]
def f() -> str: ...

[builtins fixtures/tuple.pyi]
[out]
==
main:2: error: Incompatible types in assignment (expression has type "str", variable has type "int")
```

It seems to me that Mypy does not handle this test well because of processing the modules in the right order (`a` -> `b` -> `main`) but because of storing the same `FuncDef` instance in the symbol tables of the three modules.  Hence, if `a.f` is updated when reanalysing `a`, `b` and `main` know this at once.   So, Mypy can detect the error also for the processing order `a` -> `main` -> `b` and then starts another iteration step so that the complete processing order becomes `a` -> `main` -> `b` -> `main` (it actually finds the same error twice).

Again, all this seems to apply only for `only_when_nocache`.

I suspect those familiar with Mypy's server functionalities will find it much easier to solve this puzzle (or see, where I am thinking in the wrong direction), so I will wait for help.

ilevkivskyi commented:
I will try to look at this on the weekend, there indeed may be a deeper problem.

ilevkivskyi commented:
@tyralla I think I fixed the bug in https://github.com/python/mypy/pull/17883. Please undo your `CrossRef` change. Also please add a test like this:
```
[case testDeprecateFunctionAlreadyDecorated]
from b import f
x: str = f()
import b
y: str = b.f()

[file b.py]
from a import f

[file a.py]
from typing import Callable

def d(t: Callable[[], str]) -> Callable[[], str]: ...

@d
def f() -> str: ...

[file a.py.2]
from typing import Callable
from typing_extensions import deprecated

def d(t: Callable[[], str]) -> Callable[[], str]: ...

@deprecated("deprecated decorated function")
@d
def f() -> str: ...

[builtins fixtures/tuple.pyi]
[out]
==
main:2: note: deprecated decorated function
main:4: note: deprecated decorated function
```
I am 95% sure this will fail unless you add the flag to `Decorator` snapshot as I suggested a while ago.

tyralla commented:
Okay, I'm glad to hear my analysis was correct.  Thank you for applying and merging the fix into the master branch so soon.  I reverted my fix, merged your one, and added the suggested test case `testDeprecateFunctionAlreadyDecorated`.  However, I would have been a little surprised if it failed with the current implementation because there is already a similar test (`testDeprecatedChangeFunctionDeprecationIndirectImport`) that passes.  Should I, nevertheless, add the attribute `deprecated` to the snapshot?

Due to all the refactoring and bug searching, I haven't thought about PEP 702 for a while now.   I could consider some additional special cases (you mentioned one above) and the `override` topic soon.  Do you intend to include PEP 702 support in Mypy 1.12?

ilevkivskyi commented:
Mypy always reprocesses targets with errors. So the existing test case is not testing what I want.
>  Do you intend to include PEP 702 support in Mypy 1.12?

It's up to @JukkaL, but IIUC release branch has been cut few days ago, so normally only low-risk bug fixes are accepted.

ilevkivskyi commented:
@tyralla As you see it failed as expected, so yes, you need to add `deprecated` to the snapshot. Btw I just noticed the error message I put in the test is wrong, it should start with `function a.f is deprecated:` (you can locally using `pytest -k testDeprecateFunctionAlreadyDecorated` to avoid waiting for CI).

tyralla commented:
My fault.  I did not execute the test at all because it had a slightly different prefix (`testDeprecate...` instead of `testDeprecated...`).  Too lazy...

I finally added `deprecated` to the snapshot, adjusted the test results, and split `testDeprecatedFunctionAlreadyDecorated` into two versions because of different result orders for the `cache` and the `nocache` case.

tyralla commented:
If the development of Mypy 1.12 is in such a late stage, it is probably better to wait, as this PR's title ("basic support") is really true. 

I played a little, and the first missing feature I encountered (besides the nesting cases you mentioned) is:

```python
from typing_extensions import deprecated

@deprecated("use C2 instead")
class C1: ...

def f(c: C1) -> None: ...  # no response
```

ilevkivskyi commented:
Yeah, you can create follow up PRs and/or issues.
